clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -346,8 +346,16 @@ private void cancelInFlightRequests(String nodeId, long now, Collection<ClientRe
             }
 
             if (!request.isInternalRequest) {
-                if (responses != null)
-                    responses.add(request.disconnected(now, null));
+                if (responses != null) {
+                    ClientResponse disconnected;
+
+                    if (request.timeElapsedSinceSendMs(now) > request.requestTimeoutMs)
```

Comment:
I wonder if we could pass a boolean to the method instead of re-checking `timeElapsedSinceSendMs` here. Have you tried/considered this?

Reply:
Done.